spark3/src/main/java/org/apache/iceberg/actions/Spark3MigrateAction.java

```diff
@@ -0,0 +1,138 @@
+import org.apache.iceberg.spark.SparkSessionCatalog;
+import org.apache.iceberg.spark.SparkTableUtil;
+import org.apache.iceberg.spark.source.StagedSparkTable;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.catalyst.TableIdentifier;
+import org.apache.spark.sql.connector.catalog.CatalogPlugin;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.Some;
+import scala.collection.JavaConverters;
+
+/**
+ * Takes a Spark table in the sourceCatalog and attempts to transform it into an Iceberg
+ * Table in the same location with the same identifier. Once complete the identifier which
+ * previously referred to a non-iceberg table will refer to the newly migrated iceberg
+ * table.
+ */
+class Spark3MigrateAction extends Spark3CreateAction {
+  private static final Logger LOG = LoggerFactory.getLogger(Spark3MigrateAction.class);
+  private static final String BACKUP_SUFFIX = "_BACKUP_";
+
+  Spark3MigrateAction(SparkSession spark, CatalogPlugin sourceCatalog, Identifier sourceTableName) {
+    super(spark, sourceCatalog, sourceTableName, sourceCatalog, sourceTableName);
+  }
+
+  @Override
+  public Long execute() {
+    Table icebergTable;
+
+    // Move source table to a new name, halting all modifications and allowing us to stage
+    // the creation of a new Iceberg table in its place
+    String backupName = sourceTableIdent().name() + BACKUP_SUFFIX;
+    Identifier backupIdentifier = Identifier.of(sourceTableIdent().namespace(), backupName);
+    try {
+      destCatalog().renameTable(sourceTableIdent(), backupIdentifier);
+    } catch (org.apache.spark.sql.catalyst.analysis.NoSuchTableException e) {
+      throw new NoSuchTableException("Cannot find table '%s' to migrate", sourceTableIdent());
+    } catch (org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException e) {
+      throw new AlreadyExistsException("Cannot rename migration source '%s' to backup name '%s'." +
+          " Backup table already exists.", sourceTableIdent(), backupIdentifier);
+    }
+
+    StagedSparkTable stagedTable = null;
+    boolean threw = true;
+    try {
+      stagedTable = stageDestTable();
+      icebergTable = stagedTable.table();
+
+      ensureNameMappingPresent(icebergTable);
+
+      String stagingLocation = getMetadataLocation(icebergTable);
+
+      LOG.info("Beginning migration of {} using metadata location {}", sourceTableIdent(), stagingLocation);
+      Some<String> backupNamespace = Some.apply(backupIdentifier.namespace()[0]);
+      TableIdentifier v1BackupIdentifier = new TableIdentifier(backupIdentifier.name(), backupNamespace);
+      SparkTableUtil.importSparkTable(spark(), v1BackupIdentifier, icebergTable, stagingLocation);
+
+      stagedTable.commitStagedChanges();
+      threw = false;
+    } finally {
+      if (threw) {
```

Comment:
Why use `threw` and `finally` instead of `catch`? Do we also re-throw the original exception?

Reply:
Well, it is finally, never mind.